deployment/terraform/utils.go

```diff
@@ -352,20 +355,114 @@ func getServerURL(output *Output, deploymentConfig *deployment.Config) string {
+
+	if t.config.AWSRoleARN != "" {
+		// Refresh the credentials once so that the time of the interval below
+		// starts now
+		if err := t.refreshAWSCredentialsFromRole(); err != nil {
+			return fmt.Errorf("unable to refresh token: %w", err)
+		}
+
+		// Start a goroutine refreshing the credentials every AWSRoleTokenRefreshInterval.
+		// If it fails, exit the goroutine, since it needs the previous credentials to be
+		// valid for the refresh logic to work.
+		go func() {
+			ticker := time.Tick(AWSRoleTokenRefreshInterval)
+			for range ticker {
+				if err := t.refreshAWSCredentialsFromRole(); err != nil {
+					mlog.Error("unable to refresh token, stopping goroutine", mlog.Err(err))
+					return
+				}
+			}
+		}()
+	}
 
-	if t.config.AWSProfile == "" {
-		return awsconfig.LoadDefaultConfig(
-			context.Background(),
-			regionOpt,
-		)
+	return nil
+}
+
+// refreshAWSCredentialsFromRole refreshes the current credentials by assuming
+// the role again, with an expiry duration of one hour, which is the maximum
+// allowed when using role chaining
+func (t *Terraform) refreshAWSCredentialsFromRole() error {
+	t.awsCfgMut.Lock()
+	defer t.awsCfgMut.Unlock()
+
+	if t.config.AWSRoleARN == "" {
+		return fmt.Errorf("AWSRoleARN must be non-empty")
+	}
+
+	// Test current credentials before refresh: if they are not valid, we
+	// cannot refresh
+	if _, err := t.awsCfg.Credentials.Retrieve(context.Background()); err != nil {
+		return fmt.Errorf("failed to retrieve current credentials: %w", err)
+	}
+
+	// Create new STS service and assume role again
+	stsSvc := sts.NewFromConfig(t.awsCfg)
+	creds := stscreds.NewAssumeRoleProvider(stsSvc, t.config.AWSRoleARN, func(opt *stscreds.AssumeRoleOptions) { opt.Duration = AWSRoleTokenDuration })
+	newCredCache := aws.NewCredentialsCache(creds)
+
+	// Test new credentials
+	newCreds, err := newCredCache.Retrieve(context.Background())
+	if err != nil {
+		return fmt.Errorf("failed to assume role %q with new credentials: %w", t.config.AWSRoleARN, err)
 	}
 
-	profileOpt := awsconfig.WithSharedConfigProfile(t.config.AWSProfile)
-	return awsconfig.LoadDefaultConfig(context.Background(), profileOpt, regionOpt)
+	// Update the stored configuration with the new credentials
+	mlog.Info("successfully assumed role", mlog.Time("expiry", newCreds.Expires))
+	t.awsCfg.Credentials = newCredCache
+
+	// Further verify the credentials work with a test STS call
+	if _, err := stsSvc.GetCallerIdentity(context.Background(), &sts.GetCallerIdentityInput{}); err != nil {
+		return fmt.Errorf("new credentials failed STS test: %w", err)
+	}
+
+	return nil
+}
+
+// GetAWSConfig returns the AWS config, using the profile configured in the
+// deployer if present, and defaulting to the default credential chain otherwise.
+// If a role ARN is provided, it will assume that role.
+func (t *Terraform) GetAWSConfig() (aws.Config, error) {
+	t.awsCfgMut.Lock()
+	defer t.awsCfgMut.Unlock()
+	return t.awsCfg.Copy(), nil
 }
```

Comment:
This does not seem to be called anymore. Intentional?

Reply:
Ah sorry! I thought this method was implemented new for this PR :facepalm:

Reply:
The diff UI is terrible, I don't blame you :joy: